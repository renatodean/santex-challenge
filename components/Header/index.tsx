import React from "react";
import styled from "@emotion/styled";
import Image from "next/image";
import SearchInput from "../SearchInput";
import CartIcon from "../CartIcon";
import { useRouter } from "next/router";

const HeaderContainer = styled.header<HeaderProps>`
  background-color: white;
  display: grid;
  grid-template-columns: ${({ withSearch }) =>
    withSearch ? "1fr 1fr 1fr" : "1fr 1fr"};
  height: 60px;
  padding: 10px 20px;
  box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.1);
  margin-bottom: 10px;
`;

const HomeLink = styled.div`
  cursor: pointer;
  position: relative;
  margin: 10px 0;
  height: 20px;
  width: 90px;
  @media (min-width: 420px) {
    margin: 5px 0;
    height: 30px;
    width: 155px;
  }
  @media (min-width: 1200px) {
    margin: 0;
    height: 40px;
    width: 180px;
  }
`;

interface HeaderProps {
  withSearch?: boolean;
}

const Header = ({ withSearch = false }: HeaderProps) => {
  const router = useRouter();

  return (
    <HeaderContainer withSearch={withSearch}>
      <HomeLink
        tabIndex={0}
        onClick={() => router.push("/")}
        onKeyPress={(event) => {
          if (event.key === "Enter") {
            router.push("/");
          }
        }}
        data-testid="home-link"
      >
        <Image
          priority
          src="/santex-logo.png"
          alt="Santex Logo"
          layout="fill"
        />
      </HomeLink>
      {withSearch && <SearchInput />}
      <CartIcon />
    </HeaderContainer>
  );
};

export default Header;
