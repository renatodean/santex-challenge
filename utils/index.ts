import { Item } from "../types/product";

/**
 * From a given product it will return the price as a formated string
 * @param product
 * @param quantity
 * @returns string
 */
export const getPrice = (product: Item, quantity?: number) => {
  const { variants } = product || {};
  const productPrice = variants && variants.length > 0 ? variants[0].price : 0;

  return ((productPrice * (quantity || 1)) / 100).toFixed(2);
};

/**
 * From a given product it will return the price as number
 * @param product
 * @param quantity
 * @returns number
 */
export const getPriceNumber = (product: Item, quantity?: number) => {
  const { variants } = product || {};
  const productPrice = variants && variants.length > 0 ? variants[0].price : 0;

  return (productPrice * (quantity || 1)) / 100;
};

/**
 * From a given query from router it will return the query value
 * @param query
 * @returns string
 */
export const getQueryValue = (query: any) => {
  if (Array.isArray(query)) {
    return query[0];
  } else {
    return query || "";
  }
};
