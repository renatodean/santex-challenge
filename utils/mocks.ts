export const mockedProducts = {
  products: {
    items: [
      {
        __typename: "Product",
        id: "1",
        name: "Laptop",
        slug: "laptop",
        description:
          "Now equipped with seventh-generation Intel Core processors, Laptop is snappier than ever. From daily tasks like launching apps and opening files to more advanced computing, you can power through your day thanks to faster SSDs and Turbo Boost processing up to 3.6GHz.",
        featuredAsset: {
          __typename: "Asset",
          id: "1",
          name: "derick-david-409858-unsplash.jpg",
          type: "IMAGE",
          source:
            "https://demo.vendure.io/assets/source/b6/derick-david-409858-unsplash.jpg",
          preview:
            "https://demo.vendure.io/assets/preview/71/derick-david-409858-unsplash__preview.jpg",
        },
        variants: [
          { __typename: "ProductVariant", price: 129900 },
          { __typename: "ProductVariant", price: 139900 },
          { __typename: "ProductVariant", price: 219900 },
          { __typename: "ProductVariant", price: 229900 },
        ],
      },
      {
        __typename: "Product",
        id: "2",
        name: "Tablet",
        slug: "tablet",
        description:
          'If the computer were invented today, what would it look like? It would be powerful enough for any task. So mobile you could take it everywhere. And so intuitive you could use it any way you wanted — with touch, a keyboard, or even a pencil. In other words, it wouldn’t really be a "computer." It would be Tablet.',
        featuredAsset: {
          __typename: "Asset",
          id: "2",
          name: "kelly-sikkema-685291-unsplash.jpg",
          type: "IMAGE",
          source:
            "https://demo.vendure.io/assets/source/5a/kelly-sikkema-685291-unsplash.jpg",
          preview:
            "https://demo.vendure.io/assets/preview/b8/kelly-sikkema-685291-unsplash__preview.jpg",
        },
        variants: [
          { __typename: "ProductVariant", price: 32900 },
          { __typename: "ProductVariant", price: 44500 },
        ],
      },
    ],
  },
};

export const mockedEmptyProducts = { products: { items: [] } };

export const mockedProduct = {
  product: {
    __typename: "Product",
    id: "1",
    name: "Laptop",
    slug: "laptop",
    description:
      "Now equipped with seventh-generation Intel Core processors, Laptop is snappier than ever. From daily tasks like launching apps and opening files to more advanced computing, you can power through your day thanks to faster SSDs and Turbo Boost processing up to 3.6GHz.",
    featuredAsset: {
      __typename: "Asset",
      id: "1",
      name: "derick-david-409858-unsplash.jpg",
      type: "IMAGE",
      source:
        "https://demo.vendure.io/assets/source/b6/derick-david-409858-unsplash.jpg",
      preview:
        "https://demo.vendure.io/assets/preview/71/derick-david-409858-unsplash__preview.jpg",
    },
    variants: [
      { __typename: "ProductVariant", price: 129900 },
      { __typename: "ProductVariant", price: 139900 },
      { __typename: "ProductVariant", price: 219900 },
      { __typename: "ProductVariant", price: 229900 },
    ],
  },
};

export const mockedStorageState = {
  showSnackbar: true,
  products: [
    {
      product: mockedProduct.product,
      quantity: 3,
    },
  ],
};
