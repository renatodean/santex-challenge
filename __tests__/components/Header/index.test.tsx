import { render, screen, fireEvent } from "@testing-library/react";
import Header from "../../../components/Header";
import { mockedStorageState } from "../../../utils/mocks";

const routerPushSpy = jest.fn();

jest.mock("next/router", () => ({
  useRouter() {
    return {
      route: "/",
      pathname: "",
      query: { q: "" },
      asPath: "",
      push: routerPushSpy,
      events: {
        on: jest.fn(),
        off: jest.fn(),
      },
      beforePopState: jest.fn(() => null),
      prefetch: jest.fn(() => null),
    };
  },
}));

jest.mock("../../../context", () => ({
  __esModule: true,
  useStorageContext: jest.fn(() => [mockedStorageState, jest.fn()]),
}));

describe("Header", () => {
  it("renders the header with search", () => {
    render(<Header withSearch />);

    expect(screen.getByTestId("search-input")).toBeInTheDocument();
  });

  it("should open the cart popup when icon is clicked", () => {
    render(<Header />);

    const cartIcon = screen.getByTestId("cart-icon");
    fireEvent(
      cartIcon,
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      })
    );

    expect(screen.getByTestId("cart-popup")).toBeInTheDocument();
  });

  it("should go to the homepage when press enter on page logo", () => {
    render(<Header />);

    const homeLink = screen.getByTestId("home-link");
    fireEvent.keyPress(homeLink, { key: "Enter", code: "Enter", charCode: 13 });

    expect(routerPushSpy).toHaveBeenCalled();
    expect(routerPushSpy).toHaveBeenCalledTimes(1);
    expect(routerPushSpy).toHaveBeenCalledWith("/");
  });

  it("should go to the homepage when clicks on page logo", () => {
    render(<Header />);

    const homeLink = screen.getByTestId("home-link");
    fireEvent(
      homeLink,
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      })
    );

    expect(routerPushSpy).toHaveBeenCalled();
    expect(routerPushSpy).toHaveBeenCalledTimes(2);
    expect(routerPushSpy).toHaveBeenCalledWith("/");
  });

  it("should add query when search changes", () => {
    render(<Header withSearch />);

    const homeLink = screen.getByTestId("search-input");
    fireEvent.change(homeLink, { target: { value: "Query" } });

    expect(routerPushSpy).toHaveBeenCalled();
    expect(routerPushSpy).toHaveBeenCalledTimes(3);
    expect(routerPushSpy).toHaveBeenNthCalledWith(
      3,
      {
        pathname: "/",
        query: { q: "Query" },
      },
      undefined,
      {}
    );
  });
});

describe("Header snapshot", () => {
  it("renders the header with search", () => {
    const { container } = render(<Header withSearch />);
    expect(container).toMatchSnapshot();
  });
});
