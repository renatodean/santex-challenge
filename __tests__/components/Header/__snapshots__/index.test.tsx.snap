// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header snapshot renders the header with search 1`] = `
<div>
  <header
    class="css-1npeptk"
  >
    <div
      class="css-1ui6qe5"
      data-testid="home-link"
      tabindex="0"
    >
      <span
        style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
      >
        <img
          alt="Santex Logo"
          data-nimg="fill"
          decoding="async"
          sizes="100vw"
          src="/_next/image?url=%2Fsantex-logo.png&w=3840&q=75"
          srcset="/_next/image?url=%2Fsantex-logo.png&w=640&q=75 640w, /_next/image?url=%2Fsantex-logo.png&w=750&q=75 750w, /_next/image?url=%2Fsantex-logo.png&w=828&q=75 828w, /_next/image?url=%2Fsantex-logo.png&w=1080&q=75 1080w, /_next/image?url=%2Fsantex-logo.png&w=1200&q=75 1200w, /_next/image?url=%2Fsantex-logo.png&w=1920&q=75 1920w, /_next/image?url=%2Fsantex-logo.png&w=2048&q=75 2048w, /_next/image?url=%2Fsantex-logo.png&w=3840&q=75 3840w"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
        />
      </span>
    </div>
    <div
      class="css-1g2qfl"
    >
      <input
        class="css-164cp4c"
        data-testid="search-input"
        maxlength="30"
        placeholder="Search a product!"
        value=""
      />
    </div>
    <div
      class="css-1jdhora"
    >
      <div
        class="css-piilf0"
        data-testid="cart-icon"
        tabindex="0"
      >
        <svg
          aria-hidden="true"
          class="StyledIconBase-ea9ulj-0 iLHuDN"
          fill="currentColor"
          focusable="false"
          height="30"
          viewBox="0 0 16 16"
          width="30"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 1.5A.5.5 0 0 1 .5 1H2a.5.5 0 0 1 .485.379L2.89 3H14.5a.5.5 0 0 1 .491.592l-1.5 8A.5.5 0 0 1 13 12H4a.5.5 0 0 1-.491-.408L2.01 3.607 1.61 2H.5a.5.5 0 0 1-.5-.5zM5 12a2 2 0 1 0 0 4 2 2 0 0 0 0-4zm7 0a2 2 0 1 0 0 4 2 2 0 0 0 0-4zm-7 1a1 1 0 1 1 0 2 1 1 0 0 1 0-2zm7 0a1 1 0 1 1 0 2 1 1 0 0 1 0-2z"
          />
        </svg>
        <div
          class="css-1d0bl5n"
        >
          3
        </div>
      </div>
    </div>
  </header>
</div>
`;
